### ISSUE TITLE
Font substitutions in Firestorm

### ISSUE THREAD

**Issue by ntheo1979** [lmiphay/gentoo.overlay#16]
Greetings.

I've had some progress making the unifont substitution work. I copied _unifont.ttf_ and _unifont_upper.ttf_ from /usr/share/fonts/unifont/ to /opt/firestorm-bin/fonts/ and amended /opt/firestorm-bin/fonts/fonts.xml to
```  <font name="default" comment="default font files (global fallbacks)>
    <file>DejaVuSans.ttf</file>
+   <file>unifont_upper.ttf</file>
+   <file>unifont.ttf</file>
    <os name="Windows">
      <file>meiryo.TTC</file>
⋮
```
By doing that, the viewer manages to render _Ⲙⲁⲥⳙ_ and _🌴🌞_, BUT
- only as root (not as user)
- only with unifont; trying the same trick with NotoSans-Regular and NotoColorEmoji failed.

For the second, I wonder if the viewer gets confused because it ships its own version of NotoSans-Regular.ttf (combined with NotoSansSymbols-Regular.ttf into NotoSans.ttf, it says). My intuition is to nuke them both, replace them with the proper NotoSans-Regular and NotoColorEmoji and perhaps more, and amend fonts_noto.xml thusly:
```
-    <file>NotoSans.ttf</file>
+   <file>NotoSans-Regular.ttf</file>
+   <file>NotoColorEmoji.ttf</file>
⋮
```
But I haven't tried that yet.

For the first, what gives? Should my user be in any groups that he isn't? Is there even a group regarding fonts? I've never seen this before.

**Comment by lmiphay** [comment]
tl;dr - Needs a viewer update to support color emjois. If Linden Lab uses the Catznip contribution, Firestorm will get it.

Its look the history of color emojis is something like this:

In 2016, "Colour emojis use a new font format which would require a upgrade to the font renderer." [1]
In 2019, "Emojis are coming to the viewer, courtesy of a code contribution that’s heading to Linden Lab."
         "The emoji capability is being contributed by the Catznip viewer team" [2]

Firestorm, Linden bugs in 2019 [3,4] - the Linden bug is accepted.

Links:

1. https://jira.secondlife.com/browse/BUG-40475?focusedCommentId=510128&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-510128
2. https://modemworld.me/2019/11/16/2019-tpvd-meeting-week-46-summary/
3. https://jira.firestormviewer.org/browse/FIRE-23880
4. https://jira.secondlife.com/browse/BUG-226971

**Comment by lmiphay** [comment]
I just added a -r3 ebuild - lets see how that works for now. It just adds unifont to the default fonts list.

I will have a look at the NotoSans-Regular situation a bit later.

**Comment by ntheo1979** [comment]
Thank you for your research and for your work. I unmerged -r2, deleted the fonts I had copied and merged the -r3. Again, as root I could see the substitution but not as user. Then I thought root had the default settings (therefore Dejavu as a default UI font), so I switched on my user's preferences from Noto to Dejavu and… now I can see the substitution. That means all characters and black-and-white emoji in unifont, but that's perfectly good enough for now.

Now, I only have one problem left, but that's for another day.

Again, thank you Paul.
